File: 0050-powx-n/0050-powx-n.cpp
class Solution {
public:
    double myPow(double x, int n) {
        double res = 1;
        long nn = n;
        if(n<0){
            x = 1.0/x;
            nn *= -1;
        }

        while(nn) {
            if(nn%2){
                res = res*x;
                nn -= 1;
            }
            x *= x;
            nn >>= 1;
        }
        return res;
    }

        /*
            x^n
            2^-1 = 1/2
            2^-5 = 1/2^5
        */
};

File: 0050-powx-n/0050-powx-n.py
class Solution:
    def myPow(self, x: float, n: int) -> float:
        
        if n == 0:
            return 1

        if n < 0:
            n = n*-1
            x = 1/x

        powed = 1
        while n > 0:
            if n%2:
                powed *= x
                n -= 1
            x *= x
            n //= 2

        return powed


'''
    x=3, n=5

    3*3*3*3*3

    3*3 = 9 * 9 
'''        